servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,502 @@
+   * @param journal the journal
+   */
+  public LineageMaster(FileSystemMaster fileSystemMaster, Journal journal) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("lineage-master-%d", true)));
+
+    mTachyonConf = MasterContext.getConf();
+    mFileSystemMaster = Preconditions.checkNotNull(fileSystemMaster);
+    mLineageIdGenerator = new LineageIdGenerator();
+    mLineageStore = new LineageStore(mLineageIdGenerator);
+    mWorkerToCheckpointFile = Maps.newHashMap();
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new LineageMasterService.Processor<LineageMasterServiceHandler>(
+        new LineageMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getServiceName() {
+    return Constants.LINEAGE_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    if (entry instanceof LineageEntry) {
+      mLineageStore.addLineageFromJournal((LineageEntry) entry);
+    } else if (entry instanceof LineageIdGeneratorEntry) {
+      mLineageIdGenerator.fromJournalEntry((LineageIdGeneratorEntry) entry);
+    } else if (entry instanceof AsyncCompleteFileEntry) {
+      asyncCompleteFileFromEntry((AsyncCompleteFileEntry) entry);
+    } else if (entry instanceof PersistFilesEntry) {
+      persistFilesFromEntry((PersistFilesEntry) entry);
+    } else if (entry instanceof RequestFilePersistenceEntry) {
+      requestFilePersistenceFromEntry((RequestFilePersistenceEntry) entry);
+    } else if (entry instanceof DeleteLineageEntry) {
+      deleteLineageFromEntry((DeleteLineageEntry) entry);
+    } else {
+      throw new IOException(ExceptionMessage.UNEXPECETD_JOURNAL_ENTRY.getMessage(entry));
+    }
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    super.start(isLeader);
+    if (isLeader) {
+      mCheckpointExecutionService =
+          getExecutorService().submit(new HeartbeatThread("Checkpoint scheduling service",
+              new CheckpointSchedulingExcecutor(this),
+              mTachyonConf.getInt(Constants.MASTER_LINEAGE_CHECKPOINT_INTERVAL_MS)));
+      mRecomputeExecutionService =
+          getExecutorService().submit(new HeartbeatThread("Recompute service",
+              new RecomputeExecutor(new RecomputePlanner(mLineageStore, mFileSystemMaster),
+                  mFileSystemMaster),
+              mTachyonConf.getInt(Constants.MASTER_LINEAGE_RECOMPUTE_INTERVAL_MS)));
+    }
+  }
+
+  @Override
+  public void stop() throws IOException {
+    super.stop();
+    if (mCheckpointExecutionService != null) {
+      mCheckpointExecutionService.cancel(true);
+    }
+    if (mRecomputeExecutionService != null) {
+      mRecomputeExecutionService.cancel(true);
+    }
+  }
+
+  @Override
+  public synchronized void streamToJournalCheckpoint(JournalOutputStream outputStream)
+      throws IOException {
+    mLineageStore.streamToJournalCheckpoint(outputStream);
```

Comment:
don't we need
`outputStream.writeEntry(mLineageIdGenerator.toJournalEntry());` here?